### ISSUE TITLE
feat: multi-thread rendering

### ISSUE THREAD

**Issue by UlyssesZh** [jekyll/jekyll#9485]
## Summary

Rendering is very slow for big sites or slow custom converters. However, I think it is possible to significantly speed up rendering by utilizing multi-thread.

## Motivation

I wanted to try Pandoc instead of Kramdown as the markup engine. However, every time a Markdown is processed, a new process is initiated (because pandoc is an external binary executable), not to mention it may trigger other processes (because I add custom filters to pandoc). Waiting for processes to initiate and destroy has been taking up most of the time of rendering.

However, since most of the rendering jobs do not depend on others, we may possibly alter the rendering strategy to multi-thread. In my use case, each thread will spawn their child processes concurrently, so it can be much faster by utilizing multiple cores of CPU.

## Reference-level explanation

I actually tried writing a simple plugin to let Jekyll render pages and docs concurrently:

```ruby
module Jekyll::UlyssesZhan
end

module Jekyll
	module UlyssesZhan::MultithreadRendering
		Site.prepend self

		def render *args
			@rendering_threads = []
			super
		end

		def render_regenerated *args
			@rendering_threads.push Thread.new { super }
		end

		def render_pages *args
			super
			@rendering_threads.each &:join
		end
	end
end
```

So this is just a simple try and it was partially successful. It reduced the building time from 45s to 10s, which is great. However, because the current implementation of rendering in Jekyll is not thread-safe, the rendered pages are wrong (one page has the content of another page). These could be fixed after some refactoring.

Another problem that I had is that this plugin is not compatible with the jekyll-feed plugin. It turns out that this is because feed.xml requires the posts to be rendered first so that it can contain the rendered contents. This means that we should probably have a way to indicate that some pages/docs cannot be rendered concurrently with other pages/docs. We can possibly use a front-matter entry to indicate that.

**Comment by fauno** [comment]
>  We can possibly use a front-matter entry to indicate that.

Maybe priorities, like hook plugins have?

**Comment by fauno** [comment]
> However, because the current implementation of rendering in Jekyll is not thread-safe, the rendered pages are wrong (one page has the content of another page). These could be fixed after some refactoring.

This is because the site payload used on `Site#render_regenerated` is shared, and `Renderer#assign_pages!` and others modify it.  I got better results by `dup`ing the payload on your plugin, but it's probably not the best way to do it.

**Comment by UlyssesZh** [comment]
> >  We can possibly use a front-matter entry to indicate that.
> 
> Maybe priorities, like hook plugins have?

I am thinking about something called `rendering_priority` (not just `priority` because maybe we will have other features that need specifying priorities in the future):

```yaml
---
rendering_priority: low
---
```

The rendering threads are grouped according to the priorities. Threads with lower priority is `new`ed after the group with higher priority finished `each &:join`ing.

**Comment by UlyssesZh** [comment]
So it turns out it is `Liquid::Template#render` that is not thread safe. That's why the rendered results are still incorrect even if each renderer uses a different payload. I think I may need to submit an issue on Shopify/liquid to ask about whether there is a way to let Liquid render in a thread-safe way.

For now, a workaround can be disabling the caching in `Jekyll::LiquidRenderer::File#parse` here:
https://github.com/jekyll/jekyll/blob/3f3a283018a976da11a0bfcc13a20d43d37ee29f/lib/jekyll/liquid_renderer/file.rb#L11-L18

I now basically can let Jekyll render in a multi-thread way. See [the plugin on my blog](https://github.com/UlyssesZh/UlyssesZh.github.io/blob/0b10d7fb3facdb89958040d10346f34cfb487680/_plugins/multithread_rendering.rb). It seems like a good-enough workaround.

However, this issue should still remain opening. I think multi-thread rendering should be a basic feature that users can enable without using a plugin.

**Comment by fauno** [comment]
Thanks! I'll try to run it over some big sites we have soon :)

**Comment by rouralberto** [comment]
> So it turns out it is `Liquid::Template#render` that is not thread safe. That's why the rendered results are still incorrect even if each renderer uses a different payload. I think I may need to submit an issue on Shopify/liquid (done now: [Shopify/liquid#1754](https://github.com/Shopify/liquid/issues/1754)) to ask about whether there is a way to let Liquid render in a thread-safe way.
> 
> For now, a workaround can be disabling the caching in `Jekyll::LiquidRenderer::File#parse` here:
> 
> https://github.com/jekyll/jekyll/blob/3f3a283018a976da11a0bfcc13a20d43d37ee29f/lib/jekyll/liquid_renderer/file.rb#L11-L18
> 
> I now basically can let Jekyll render in a multi-thread way. See [the plugin on my blog](https://github.com/UlyssesZh/UlyssesZh.github.io/blob/b30abe0442f796c95ae239917b08b000a34df1ae/_plugins/multithread_rendering.rb#L3). It seems like a good-enough workaround.
> 
> Even though workaround is possible, this issue should remain open. I think multi-thread rendering should be a basic feature that users can enable without using a plugin.

I tried to use your plugin in my site but it gets stuck in `Generating...`.

--

I'm using Jekyll 4.3.2. Any pointers on how to make it work? Did you have any similar issue while developing it?

**Comment by mattr-** [comment]
> Any pointers on how to make it work? Did you have any similar issue while developing it?

Please take any issues with a particular plugin to that plugin's repo. Thanks! ❤

**Comment by fauno** [comment]
@mattr- any plans for this? maybe #9156 ?

**Comment by mattr-** [comment]
Maybe. I'm not super focused on Jekyll 5 at the moment, but rather a 4.4 release. Since this issue is already linked there (along with a comment on that issue requesting multi-core support), I think that will cover everything.

**Comment by UlyssesZh** [comment]
> Please take any issues with a particular plugin to that plugin's repo. Thanks! ❤

Yes, I do not want to deviate the discussion in this issue either, but that plugin does not have its own repo. I created UlyssesZh/UlyssesZh.github.io#95 for discussions specifically for that plugin. @rouralberto Please continue there.

For other people watching this issue, if you want to talk about the workaround plugin I wrote and linked, please go to UlyssesZh/UlyssesZh.github.io#95.
